# iNaturalist status updates by state

Using the file produced in the collate-status-taxa.ipynb: `inat-aust-status-taxa.csv`, generate lists to update iNaturalist statuses

**Next steps:**
State by state establish the changes that need to be made:
    a. new - any new species that appear in the state lists but do not have a status in inaturalist (new template)
    b. updates - any changes to information which was added by us previously (user_id = 708886) (update template, action='UPDATE')
    c. removals - any statuses which were added by us previously (user_id = 708886) list which are incorrect (update template, action='REMOVE')
    d. flags - are there any statuses by other users that need to be flagged?

## Prep - common to all states
1. Read in the inaturalist statuses & filter out this state
2. Read in the inaturalist taxa list
3. Read in the state sensitive list
4. Attempt to match the state statuses to an IUCN equivalent


### 1. iNaturalist statuses

In [28]:
import pandas as pd

# projectdir = "/Users/oco115/PycharmProjects/authoritative-lists/" # basedir for this gh project
projectdir = "/Users/new330/IdeaProjects/authoritative-lists/" # basedir for this gh project
sourcedir = projectdir + "source-data/inaturalist-statuses/"
listdir = projectdir + "current-lists/"


# read in the statuses
taxastatus = pd.read_csv(sourcedir + "inat-aust-status-taxa.csv", encoding='UTF-8',na_filter=False,dtype=str) ## Read inaturalist conservation statuses file
taxastatus.head(3)

id taxon_id  user_id place_id source_id  \
0  166449    38493  1138587     7830             
1  234788   918383   702203     9994             
2  234789   918383   702203     7308             

                            authority status  \
0  Flora and Fauna Guarantee Act 1988     CR   
1           Atlas of Living Australia    NT    
2           Atlas of Living Australia     LC   

                                                 url description geoprivacy  \
0                                                                  obscured   
1  https://bie.ala.org.au/species/https://id.biod...                          
2  https://bie.ala.org.au/species/https://id.biod...                          

   ...         genus specificEpithet infraspecificEpithet  \
0  ...     Eulamprus      kosciuskoi                        
1  ...  Chiloschista     phyllorhiza                        
2  ...  Chiloschista     phyllorhiza                        

               modified            scientificName taxonRank  \
0  2021-03-01T10:35:01Z      Eulamprus kosciuskoi   species   
1  2022-01-08T03:30:36Z  Chiloschista phyllorhiza   species   
2  2022-01-08T03:30:36Z  Chiloschista phyllorhiza   species   

                                          references preferred_common_name  \
0  http://reptile-database.reptarium.cz/search.ph...                         
1  http://www.catalogueoflife.org/annual-checklis...                         
2  http://www.catalogueoflife.org/annual-checklis...                         

  is_active current_synonymous_taxon_ids  
0                                         
1                                         
2                                         

[3 rows x 34 columns]

In [30]:
def filter_state_statuses(stateregex: str, urlregex: str):
    authoritydf = taxastatus['authority'].drop_duplicates().sort_values()
    authoritydf = authoritydf[pd.Series(authoritydf).str.contains(stateregex)]
    urldf = taxastatus['url'].drop_duplicates().sort_values()
    urldf = urldf[pd.Series(urldf).str.contains(urlregex)]
    placedisplaydf = taxastatus['place_display_name'].drop_duplicates().sort_values()
    placedisplaydf = placedisplaydf[pd.Series(placedisplaydf).str.contains(stateregex)]
    placedf = taxastatus['place_name'].drop_duplicates().sort_values()
    placedf = placedf[pd.Series(placedf).str.contains(stateregex)]
    # concat all and remove duplicates
    statedf = pd.concat([taxastatus.apply(lambda row: row[taxastatus['place_display_name'].isin(placedisplaydf)]),
                         taxastatus.apply(lambda row: row[taxastatus['place_name'].isin(placedf)]),
                         taxastatus.apply(lambda row: row[taxastatus['url'].isin(urldf)]),
                         taxastatus.apply(
                             lambda row: row[taxastatus['authority'].isin(authoritydf)])]).drop_duplicates()
    return statedf.sort_values(['taxon_id', 'user_id'])


inatstatuses = filter_state_statuses("Qld|QLD|Queensland|QUEENSLAND|QL", ".qld.")
inatstatuses.rename(columns={'id':'status_id','id_y':'taxon_id_y'},inplace=True)
inatstatuses

status_id taxon_id  user_id place_id source_id  \
532     223098  1003032  3669610     7308             
131     169944  1004434    58320     7308             
1473    152757   100615   708886     7308     16653   
1407    152668   100635   708886     7308     16653   
1267    152461   100638   708886     7308     16653   
...        ...      ...      ...      ...       ...   
835     262047   977629  3669610     7308             
2615    161807   993333   702203     7308             
2616    161808   993333   702203     6744             
23      167723   993605  3669610     7308             
2897    164338   993923    58320     7308             

                                              authority           status  \
532             Queensland Nature Conservation Act 1992               CE   
131             Queensland Nature Conservation Act 1992    Least concern   
1473                                           QLD DEHP       endangered   
1407                                           QLD DEHP       endangered   
1267                                           QLD DEHP       endangered   
...                                                 ...              ...   
835             Queensland Nature Conservation Act 1992               NT   
2615                      Nature Conservation Act 1992                VU   
2616  Environment Protection and Biodiversity Conser...               VU   
23                                             QLD DEHP               NT   
2897            Queensland Nature Conservation Act 1992  Near threatened   

                                                    url  \
532   https://apps.des.qld.gov.au/species-search/det...   
131   https://apps.des.qld.gov.au/species-search/det...   
1473  https://data.qld.gov.au/dataset/conservation-s...   
1407  https://data.qld.gov.au/dataset/conservation-s...   
1267  https://data.qld.gov.au/dataset/conservation-s...   
...                                                 ...   
835   https://www.data.qld.gov.au/dataset/conservati...   
2615  https://apps.des.qld.gov.au/species-search/det...   
2616  https://apps.des.qld.gov.au/species-search/det...   
23    https://www.data.qld.gov.au/dataset/conservati...   
2897  https://apps.des.qld.gov.au/species-search/det...   

                                            description geoprivacy  ...  \
532                                                       obscured  ...   
131   Listed as 'confidential' by the Queensland Dep...   obscured  ...   
1473                                                      obscured  ...   
1407                                                      obscured  ...   
1267                                                      obscured  ...   
...                                                 ...        ...  ...   
835                                                           open  ...   
2615                                                                ...   
2616                                                                ...   
23                                                        obscured  ...   
2897       See https://www.inaturalist.org/flags/512314       open  ...   

            genus       specificEpithet infraspecificEpithet  \
532    Antrophyum  austroqueenslandicum                        
131      Oberonia         crateriformis                        
1473    Euastacus                bindal                        
1407    Euastacus                jagara                        
1267    Euastacus                maidae                        
...           ...                   ...                  ...   
835     Melaleuca               formosa                        
2615  Cupaniopsis            tomentella                        
2616  Cupaniopsis            tomentella                        
23      Acianthus         amplexicaulis                        
2897      Ardisia                bakeri                        

                  modified                   scienti

### 2. iNaturalist taxonomy

In [3]:
# Output files contain these fields
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# so we need to match species from the state lists to the inat taxa to get the taxon_id

import zipfile
url = "https://www.inaturalist.org/taxa/inaturalist-taxonomy.dwca.zip"
filename = url.split("/")[-1]

z=zipfile.ZipFile(sourcedir + filename)

with z.open('taxa.csv') as from_archive:
    inattaxa = pd.read_csv(from_archive,dtype=str)
z.close()
inattaxa.head(3)


id                             taxonID                          identifier  \
0  1  https://www.inaturalist.org/taxa/1  https://www.inaturalist.org/taxa/1   
1  2  https://www.inaturalist.org/taxa/2  https://www.inaturalist.org/taxa/2   
2  3  https://www.inaturalist.org/taxa/3  https://www.inaturalist.org/taxa/3   

                         parentNameUsageID   kingdom    phylum class order  \
0   https://www.inaturalist.org/taxa/48460  Animalia       NaN   NaN   NaN   
1       https://www.inaturalist.org/taxa/1  Animalia  Chordata   NaN   NaN   
2  https://www.inaturalist.org/taxa/355675  Animalia  Chordata  Aves   NaN   

  family genus specificEpithet infraspecificEpithet              modified  \
0    NaN   NaN             NaN                  NaN  2021-11-02T06:05:44Z   
1    NaN   NaN             NaN                  NaN  2021-11-23T00:40:18Z   
2    NaN   NaN             NaN                  NaN  2022-12-27T07:33:16Z   

  scientificName taxonRank                                         references  
0       Animalia   kingdom  http://www.catalogueoflife.org/annual-checklis...  
1       Chordata    phylum  http://www.catalogueoflife.org/annual-checklis...  
2           Aves     class  http://www.catalogueoflife.org/annual-checklis...

### 3. State lists

In [4]:
sensitivelist = pd.read_csv(listdir + "sensitive-lists/QLD-sensitive.csv")  # Qld sensitive list
sensitivelist['scientificName'] = sensitivelist['scientificName'].str.replace('subsp. ', '', regex=False)
sensitivelist = sensitivelist.rename(columns={'taxonID':'wildnetTaxonID'})
sensitivelist

wildnetTaxonID   kingdom          class            family  \
0               969  Animalia       Mammalia     Rhinolophidae   
1              1376  Animalia           Aves       Estrildidae   
2              1378  Animalia           Aves       Estrildidae   
3              1370  Animalia           Aves       Estrildidae   
4              1365  Animalia           Aves       Estrildidae   
..              ...       ...            ...               ...   
950           11699   Plantae  Equisetopsida  Thelypteridaceae   
951           11700   Plantae  Equisetopsida  Thelypteridaceae   
952           16042   Plantae  Equisetopsida  Thelypteridaceae   
953            8185   Plantae  Equisetopsida        Proteaceae   
954            8486   Plantae  Equisetopsida       Psilotaceae   

                   scientificName  \
0      Rhinolophus philippinensis   
1               Chloebia gouldiae   
2              Erythrura trichroa   
3    Neochmia phaeton evangelinae   
4          Poephila cincta cincta   
..                            ...   
950        Pneumatopteris costata   
951      Pneumatopteris pennigera   
952         Thelypteris confluens   
953            Macadamia jansenii   
954        Tmesipteris lanceolata   

                                     vernacularName  \
0                 greater large-eared horseshoe bat   
1                                    Gouldian finch   
2                           blue-faced parrot-finch   
3          crimson finch (white-bellied subspecies)   
4    black-throated finch (white-rumped subspecies)   
..                                              ...   
950                                             NaN   
951                                       lime fern   
952                                             NaN   
953                                             NaN   
954                                             NaN   

        scientificNameAuthorship sourceStatus Significant  \
0               Waterhouse, 1843            E           Y   
1                  (Gould, 1844)            E           Y   
2               (Kittlitz, 1835)           NT           Y   
3    (Hombron & Jacquinot, 1841)            E           Y   
4                    Gould, 1837            E           Y   
..                           ...          ...         ...   
950             (Brack.) Holttum           NT           Y   
951           (G.Forst.) Holttum            E           Y   
952          (Thunb.) C.V.Morton            V           Y   
953       C.L.Gross & P.H.Weston           CR           Y   
954                    P.A.Dang.           PE           Y   

                    status                Endemicity EPBC Status  
0               Endangered          Regional Endemic  Vulnerable  
1               Endangered             Intranational  Endangered  
2          Near Threatened  Not Endemic to Australia         NaN  
3               Endangered          Regional Endemic  Endangered  
4               Endangered             Intranational  Endangered  
..                     ...                       ...         ...  
950        Near Threatened          Regional Endemic         NaN  
951             Endangered  Not Endemic to Australia         NaN  
952             Vulnerable  Not Endemic to Australia         NaN  
953  Critically Endangered        Queensland Endemic  Endangered  
954    Extinct in the wild        Queensland Endemic     Extinct  

[955 rows x 12 columns]

### 4. Equivalent IUCN statuses

In [5]:
iucn_statuses = {'Not Evaluated', 'Data Deficient', 'Least Concern', 'Near Threatened', 'Vulnerable', 'Endangered', 'Critically Endangered', 'Extinct in the Wild' and 'Extinct'}
sensitivelist.groupby(['status'])['status'].count()

status
Critically Endangered     35
Endangered                78
Extinct in the wild        7
Least concern            143
Near Threatened           63
Special least concern    361
Vulnerable               139
Name: status, dtype: int64

In [6]:
iucnStatusMappings = {
    'critically endangered': 'Critically Endangered',
    'vulnerable':'Vulnerable',
    'not evaluated':'Not Evaluated',
    'data deficient':'Data Deficient',
    'least concern':'Least Concern',
    'special least concern':'Least Concern',
    'near threatened':'Near Threatened',
    'endangered':'Endangered',
    'extinct in the wild':'Extinct in the Wild',
    'extinct':'Extinct',
    'confidential':'Vulnerable'
}

### 5. Determine best place ID to use

In [7]:
inatstatuses.groupby(['place_id','place_name','place_display_name'])['place_id'].count()
# looks like 7308

place_id  place_name             place_display_name           
                                                                    5
144315    Brisbane City          Brisbane City                      2
153119    South East Queensland  South East Queensland, QL, AU      1
18870     Cairns - Pt B          Cairns - Pt B, QL, AU              1
19232     Yarrabah               Yarrabah, QL, AU                   1
6744      Australia              Australia                         12
7308      Queensland             Queensland, AU                   631
Name: place_id, dtype: int64

## Merge iNaturalist statuses with State sensitive list on scientificName

1. Match - updates, even if the statuses are the same we'll update the links and values anyway
2. No match - statuses to be added (additions)
   1.1 No match and no taxnomy - search for synonyms
   1.2 No match
3. Merge the other direction to see if there are deletes?


In [54]:
# join to see which lists already have a status in inaturalist based on scientificName
mergedstatuses = sensitivelist[['wildnetTaxonID','scientificName','status']].merge(inatstatuses[['status_id','scientificName','taxon_id','user_id','description','iucn','authority','status','geoprivacy','place_id','place_display_name']],how="left",left_on='scientificName',right_on='scientificName',suffixes=(None,'_inat')).sort_values(['scientificName'])
mergedstatuses


wildnetTaxonID        scientificName                 status status_id  \
127            9655             Acianthus          Least concern       NaN   
128            7974    Acianthus borealis  Special least concern       NaN   
129           13477    Acianthus caudatus  Special least concern    152636   
130           14086    Acianthus exsertus  Special least concern    152637   
131           14087  Acianthus fornicatus  Special least concern       NaN   
..              ...                   ...                    ...       ...   
126            8539    Wodyetia bifurcata             Vulnerable    152450   
55            22659     Wollumbinia belli             Vulnerable       NaN   
778           21765               Zeuxine                    NaN       NaN   
779           40091     Zeuxine attenuata          Least concern       NaN   
780           15903       Zeuxine oblonga  Special least concern    152829   

    taxon_id user_id description iucn authority status_inat geoprivacy  \
127      NaN     NaN         NaN  NaN       NaN         NaN        NaN   
128      NaN     NaN         NaN  NaN       NaN         NaN        NaN   
129   320644  708886               40  QLD DEHP  endangered   obscured   
130   323899  708886               10  QLD DEHP          LC              
131      NaN     NaN         NaN  NaN       NaN         NaN        NaN   
..       ...     ...         ...  ...       ...         ...        ...   
126   135348  708886               40  QLD DEHP  endangered   obscured   
55       NaN     NaN         NaN  NaN       NaN         NaN        NaN   
778      NaN     NaN         NaN  NaN       NaN         NaN        NaN   
779      NaN     NaN         NaN  NaN       NaN         NaN        NaN   
780   369267  708886               10  QLD DEHP          LC              

    place_id place_display_name  
127      NaN                NaN  
128      NaN                NaN  
129     7308     Queensland, AU  
130     7308     Queensland, AU  
131      NaN                NaN  
..       ...                ...  
126     7308     Queensland, AU  
55       NaN                NaN  
778      NaN                NaN  
779      NaN                NaN  
780     7308     Queensland, AU  

[958 rows x 13 columns]

In [55]:
# prepare the export fields, common to New template and Update template
# new statuses
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username
mergedstatuses['new_authority'] = "Qld Department of Environment and Science"
mergedstatuses['new_description'] = "Listed as Confidential - refer to https://www.data.qld.gov.au/dataset/queensland-confidential-species"
mergedstatuses['new_url'] = "https://apps.des.qld.gov.au/species-search/details/?id=" + mergedstatuses['wildnetTaxonID'].astype(str)
mergedstatuses['new_geoprivacy'] = "obscured"
mergedstatuses['new_place_id'] = '7308'  # Queensland, AU
mergedstatuses['new_username'] = 'peggydnew'
mergedstatuses['new_iucn_equivalent'] = mergedstatuses['status'].str.lower().str.strip().map(iucnStatusMappings).fillna('Vulnerable') # map to dictionary
mergedstatuses['new_status'] = mergedstatuses['status'].fillna('Confidential')
mergedstatuses

wildnetTaxonID        scientificName                 status status_id  \
127            9655             Acianthus          Least concern       NaN   
128            7974    Acianthus borealis  Special least concern       NaN   
129           13477    Acianthus caudatus  Special least concern    152636   
130           14086    Acianthus exsertus  Special least concern    152637   
131           14087  Acianthus fornicatus  Special least concern       NaN   
..              ...                   ...                    ...       ...   
126            8539    Wodyetia bifurcata             Vulnerable    152450   
55            22659     Wollumbinia belli             Vulnerable       NaN   
778           21765               Zeuxine                    NaN       NaN   
779           40091     Zeuxine attenuata          Least concern       NaN   
780           15903       Zeuxine oblonga  Special least concern    152829   

    taxon_id user_id description iucn authority status_inat  ... place_id  \
127      NaN     NaN         NaN  NaN       NaN         NaN  ...      NaN   
128      NaN     NaN         NaN  NaN       NaN         NaN  ...      NaN   
129   320644  708886               40  QLD DEHP  endangered  ...     7308   
130   323899  708886               10  QLD DEHP          LC  ...     7308   
131      NaN     NaN         NaN  NaN       NaN         NaN  ...      NaN   
..       ...     ...         ...  ...       ...         ...  ...      ...   
126   135348  708886               40  QLD DEHP  endangered  ...     7308   
55       NaN     NaN         NaN  NaN       NaN         NaN  ...      NaN   
778      NaN     NaN         NaN  NaN       NaN         NaN  ...      NaN   
779      NaN     NaN         NaN  NaN       NaN         NaN  ...      NaN   
780   369267  708886               10  QLD DEHP          LC  ...     7308   

    place_display_name                              new_authority  \
127                NaN  Qld Department of Environment and Science   
128                NaN  Qld Department of Environment and Science   
129     Queensland, AU  Qld Department of Environment and Science   
130     Queensland, AU  Qld Department of Environment and Science   
131                NaN  Qld Department of Environment and Science   
..                 ...                                        ...   
126     Queensland, AU  Qld Department of Environment and Science   
55                 NaN  Qld Department of Environment and Science   
778                NaN  Qld Department of Environment and Science   
779                NaN  Qld Department of Environment and Science   
780     Queensland, AU  Qld Department of Environment and Science   

                                       new_description  \
127  Listed as Confidential - refer to https://www....   
128  Listed as Confidential - refer to https://www....   
129  Listed as Confidential - refer to https://www....   
130  Listed as Confidential - refer to https://www....   
131  Listed as Confidential - refer to https://www....   
..                                                 ...   
126  Listed as Confidential - refer to https://www....   
55   Listed as Confidential - refer to https://www....   
778  Listed as Confidential - refer to https://www....   
779  Listed as Confidential - refer to https://www....   
780  Listed as Confidential - refer to https://www....   

                                               new_url new_geoprivacy  \
127  https://apps.des.qld.gov.au/species-search/det...       obscured   
128  https://apps.des.qld.gov.au/species-search/det...       obscured   
129  https://apps.des.qld.gov.au/species-search/det...       obscured   
130  https://apps.des.qld.gov.au/species-search/det...       obscured   
131  https://apps.des.qld.gov.au/species-search/det...       obscured   
..                                                 ...            ...   
126  https://apps.des.qld.gov.au/species-search/det...       obscured   
55   https://apps.des.qld.gov.au/

## Updates

In [73]:
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username,description
updates = pd.DataFrame(mergedstatuses[mergedstatuses['status_id'].notnull()])
updates.sort_values('scientificName')
updates['action'] = 'UPDATE'
#updates.loc[:,'action'] = 'UPDATE'
updates = updates[['action','scientificName','status_id','taxon_id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
updates.columns = updates.columns.str.replace("new_", "", regex=True)
updates = updates.rename(columns={'scientificName':'taxon_name',
                                  'status_id':'id'})
updates

action           taxon_name      id taxon_id                 status  \
129  UPDATE   Acianthus caudatus  152636   320644  Special least concern   
130  UPDATE   Acianthus exsertus  152637   323899  Special least concern   
133  UPDATE   Acianthus pusillus  152632   549968  Special least concern   
134  UPDATE  Acianthus saxatilis  152567   739107             Endangered   
135  UPDATE            Acriopsis  152867   425476          Least concern   
..      ...                  ...     ...      ...                    ...   
773  UPDATE        Vanda hindsii  152654   140333             Vulnerable   
39   UPDATE    Varanus semiremex  152553    39474          Least concern   
776  UPDATE          Vrydagzynea  152846   426364           Confidential   
126  UPDATE   Wodyetia bifurcata  152450   135348             Vulnerable   
780  UPDATE      Zeuxine oblonga  152829   369267  Special least concern   

    iucn_equivalent                                  authority  \
129   Least Concern  Qld Department of Environment and Science   
130   Least Concern  Qld Department of Environment and Science   
133   Least Concern  Qld Department of Environment and Science   
134      Endangered  Qld Department of Environment and Science   
135   Least Concern  Qld Department of Environment and Science   
..              ...                                        ...   
773      Vulnerable  Qld Department of Environment and Science   
39    Least Concern  Qld Department of Environment and Science   
776      Vulnerable  Qld Department of Environment and Science   
126      Vulnerable  Qld Department of Environment and Science   
780   Least Concern  Qld Department of Environment and Science   

                                                   url geoprivacy place_id  \
129  https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
130  https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
133  https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
134  https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
135  https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
..                                                 ...        ...      ...   
773  https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
39   https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
776  https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
126  https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
780  https://apps.des.qld.gov.au/species-search/det...   obscured     7308   

      username                                        description  
129  peggydnew  Listed as Confidential - refer to https://www....  
130  peggydnew  Listed as Confidential - refer to https://www....  
133  peggydnew  Listed as Confidential - refer to https://www....  
134  peggydnew  Listed as Confidential - refer to https://www....  
135  peggydnew  Listed as Confidential - refer to https://www....  
..         ...                                                ...  
773  peggydnew  Listed as Confidential - refer to https://www....  
39   peggydnew  Listed as Confidential - refer to https://www....  
776  peggydnew  Listed as Confidential - refer to https://www....  
126  peggydnew  Listed as Confidential - refer to https://www....  
780  peggydnew  Listed as Confidential - refer to https://www....  

[426 rows x 12 columns]

In [59]:
# investigation - which updates are mine (346), which are those not from me (80 not from me)
#updates[updates['user_id']=='708886'][['scientificName','new_status','status_inat','authority','new_authority','description','new_description','geoprivacy','new_geoprivacy']]
#updates[updates['user_id']!='708886'][['user_id','scientificName','new_status','status_inat','authority','new_authority','description','new_description','geoprivacy','new_geoprivacy']]
# those with different statuses
#updates[updates['new_status'].str.lower().str.strip() != updates['status_inat'].str.lower().str.strip()][['scientificName','new_status','status_inat','authority','new_authority','description','new_description','geoprivacy','new_geoprivacy']]
# users who've updated qld statuses who aren't me
#'https://www.inaturalist.org/users/220795','Steven Kurniawidjaja','neontetraploid','US'
#'https://www.inaturalist.org/users/3669610','Craig Robbins','craig-r','AU'
#'https://www.inaturalist.org/users/527710','James Kameron Mitchell','jameskm','US'
#'https://www.inaturalist.org/users/58320','lwnrngr','lwnrngr','NZ'
#'https://www.inaturalist.org/users/702203','Kitty Maurey','kitty12','CA'
#'https://www.inaturalist.org/users/717122','Miguel de Salas','mftasp','TAS'


scientificName             new_status    status_inat  \
129   Acianthus caudatus  Special least concern     endangered   
130   Acianthus exsertus  Special least concern             LC   
133   Acianthus pusillus  Special least concern             LC   
134  Acianthus saxatilis             Endangered     endangered   
135            Acriopsis          Least concern  Least concern   
..                   ...                    ...            ...   
773        Vanda hindsii             Vulnerable     Vulnerable   
39     Varanus semiremex          Least concern     endangered   
776          Vrydagzynea           Confidential   Confidential   
126   Wodyetia bifurcata             Vulnerable     endangered   
780      Zeuxine oblonga  Special least concern             LC   

                                   authority  \
129                                 QLD DEHP   
130                                 QLD DEHP   
133                                 QLD DEHP   
134                                 QLD DEHP   
135  Queensland Nature Conservation Act 1992   
..                                       ...   
773  Queensland Nature Conservation Act 1992   
39                                  QLD DEHP   
776                    Queensland Government   
126                                 QLD DEHP   
780                                 QLD DEHP   

                                 new_authority  \
129  Qld Department of Environment and Science   
130  Qld Department of Environment and Science   
133  Qld Department of Environment and Science   
134  Qld Department of Environment and Science   
135  Qld Department of Environment and Science   
..                                         ...   
773  Qld Department of Environment and Science   
39   Qld Department of Environment and Science   
776  Qld Department of Environment and Science   
126  Qld Department of Environment and Science   
780  Qld Department of Environment and Science   

                                           description  \
129                                                      
130                                                      
133  Least concern (Nature Conservation Act 1992) b...   
134                                                      
135  Listed as 'confidential' by the Queensland Dep...   
..                                                 ...   
773                                                      
39                                                       
776  Listed as 'confidential' by the Queensland Dep...   
126                                                      
780                                                      

                                       new_description geoprivacy  \
129  Listed as Confidential - refer to https://www....   obscured   
130  Listed as Confidential - refer to https://www....              
133  Listed as Confidential - refer to https://www....   obscured   
134  Listed as Confidential - refer to https://www....   obscured   
135  Listed as Confidential - refer to https://www....   obscured   
..                                                 ...        ...   
773  Listed as Confidential - refer to https://www....   obscured   
39   Listed as Confidential - refer to https://www....   obscured   
776  Listed as Confidential - refer to https://www....   obscured   
126  Listed as Confidential - refer to https://www....   obscured   
780  Listed as Confidential - refer to https://www....              

    new_geoprivacy  
129       obscured  
130       obscured  
133       obscured  
134       obscured  
135       obscured  
..             ...  
773       obscured  
39        obscured  
776       obscured  
126       obscured  
780       obscured  

[346 rows x 9 columns]

## No status in iNaturalist via straight scientificName match
The Qld records that didn't match up to a status in iNaturalist

In [81]:
# to add: those that have no inaturalist status - 532!!
noinatstatus = mergedstatuses[mergedstatuses['status_id'].isnull()]
# try to match the taxon name to something in inaturalist
noinatstatus = noinatstatus.merge(inattaxa, how="left", left_on="scientificName",right_on="scientificName")
noinatstatus

wildnetTaxonID        scientificName                 status status_id  \
0              9655             Acianthus          Least concern       NaN   
1              7974    Acianthus borealis  Special least concern       NaN   
2             14087  Acianthus fornicatus  Special least concern       NaN   
3              8545   Acianthus ledwardii  Special least concern       NaN   
4             31909  Acriopsis emarginata             Vulnerable       NaN   
..              ...                   ...                    ...       ...   
532           31261    Vanilla planifolia                    NaN       NaN   
533           30794     Vrydagzynea grayi             Endangered       NaN   
534           22659     Wollumbinia belli             Vulnerable       NaN   
535           21765               Zeuxine                    NaN       NaN   
536           40091     Zeuxine attenuata          Least concern       NaN   

    taxon_id user_id description iucn authority status_inat  ...  \
0        NaN     NaN         NaN  NaN       NaN         NaN  ...   
1        NaN     NaN         NaN  NaN       NaN         NaN  ...   
2        NaN     NaN         NaN  NaN       NaN         NaN  ...   
3        NaN     NaN         NaN  NaN       NaN         NaN  ...   
4        NaN     NaN         NaN  NaN       NaN         NaN  ...   
..       ...     ...         ...  ...       ...         ...  ...   
532      NaN     NaN         NaN  NaN       NaN         NaN  ...   
533      NaN     NaN         NaN  NaN       NaN         NaN  ...   
534      NaN     NaN         NaN  NaN       NaN         NaN  ...   
535      NaN     NaN         NaN  NaN       NaN         NaN  ...   
536      NaN     NaN         NaN  NaN       NaN         NaN  ...   

           phylum       class        order       family      genus  \
0    Tracheophyta  Liliopsida  Asparagales  Orchidaceae  Acianthus   
1             NaN         NaN          NaN          NaN        NaN   
2    Tracheophyta  Liliopsida  Asparagales  Orchidaceae  Acianthus   
3    Tracheophyta  Liliopsida  Asparagales  Orchidaceae  Acianthus   
4    Tracheophyta  Liliopsida  Asparagales  Orchidaceae  Acriopsis   
..            ...         ...          ...          ...        ...   
532  Tracheophyta  Liliopsida  Asparagales  Orchidaceae    Vanilla   
533           NaN         NaN          NaN          NaN        NaN   
534           NaN         NaN          NaN          NaN        NaN   
535  Tracheophyta  Liliopsida  Asparagales  Orchidaceae    Zeuxine   
536           NaN         NaN          NaN          NaN        NaN   

    specificEpithet infraspecificEpithet              modified taxonRank  \
0               NaN                  NaN  2021-04-17T03:13:28Z     genus   
1               NaN                  NaN                   NaN       NaN   
2        fornicatus                  NaN  2021-05-20T01:27:37Z   species   
3         ledwardii                  NaN  2021-05-18T04:43:56Z   species   
4        emarginata                  NaN  2020-11-16T22:24:21Z   species   
..              ...                  ...                   ...       ...   
532      planifolia                  NaN  2019-11-23T05:28:46Z   species   
533             NaN                  NaN                   NaN       NaN   
534             NaN                  NaN                   NaN       NaN   
535             NaN                  NaN  2022-10-17T14:37:53Z     genus   
536             NaN                  NaN                   NaN       NaN   

                                            references  
0    http://www.catalogueoflife.org/annual-checklis...  
1                                                  NaN  
2    http://www.catalogueoflife.org/annual-checklis...  
3    http://www.catalogueoflife.org/annual-checklis...  
4    http://www.catalogueoflife.org/annual-checklis...  
..                                                 ...  
532  http://www.catalogueoflife.org/annual-checklis...  
533                                      

In [91]:
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions
# note: "Dendrobium" matches to both genus and section

wildnetTaxonID          scientificName                 status status_id  \
0              9655               Acianthus          Least concern       NaN   
2             14087    Acianthus fornicatus  Special least concern       NaN   
3              8545     Acianthus ledwardii  Special least concern       NaN   
4             31909    Acriopsis emarginata             Vulnerable       NaN   
6             41744  Alsophila cunninghamii        Near Threatened       NaN   
..              ...                     ...                    ...       ...   
523           21627            Thrixspermum                    NaN       NaN   
525           21895                Trachoma                    NaN       NaN   
530            1093           Turnix olivii  Critically Endangered       NaN   
532           31261      Vanilla planifolia                    NaN       NaN   
535           21765                 Zeuxine                    NaN       NaN   

    taxon_id user_id description iucn authority status_inat  ...  \
0        NaN     NaN         NaN  NaN       NaN         NaN  ...   
2        NaN     NaN         NaN  NaN       NaN         NaN  ...   
3        NaN     NaN         NaN  NaN       NaN         NaN  ...   
4        NaN     NaN         NaN  NaN       NaN         NaN  ...   
6        NaN     NaN         NaN  NaN       NaN         NaN  ...   
..       ...     ...         ...  ...       ...         ...  ...   
523      NaN     NaN         NaN  NaN       NaN         NaN  ...   
525      NaN     NaN         NaN  NaN       NaN         NaN  ...   
530      NaN     NaN         NaN  NaN       NaN         NaN  ...   
532      NaN     NaN         NaN  NaN       NaN         NaN  ...   
535      NaN     NaN         NaN  NaN       NaN         NaN  ...   

           phylum           class            order       family         genus  \
0    Tracheophyta      Liliopsida      Asparagales  Orchidaceae     Acianthus   
2    Tracheophyta      Liliopsida      Asparagales  Orchidaceae     Acianthus   
3    Tracheophyta      Liliopsida      Asparagales  Orchidaceae     Acianthus   
4    Tracheophyta      Liliopsida      Asparagales  Orchidaceae     Acriopsis   
6    Tracheophyta  Polypodiopsida       Cyatheales  Cyatheaceae     Alsophila   
..            ...             ...              ...          ...           ...   
523  Tracheophyta      Liliopsida      Asparagales  Orchidaceae  Thrixspermum   
525  Tracheophyta      Liliopsida      Asparagales  Orchidaceae      Trachoma   
530      Chordata            Aves  Charadriiformes   Turnicidae        Turnix   
532  Tracheophyta      Liliopsida      Asparagales  Orchidaceae       Vanilla   
535  Tracheophyta      Liliopsida      Asparagales  Orchidaceae       Zeuxine   

    specificEpithet infraspecificEpithet              modified taxonRank  \
0               NaN                  NaN  2021-04-17T03:13:28Z     genus   
2        fornicatus                  NaN  2021-05-20T01:27:37Z   species   
3         ledwardii                  NaN  2021-05-18T04:43:56Z   species   
4        emarginata                  NaN  2020-11-16T22:24:21Z   species   
6      cunninghamii                  NaN  2019-02-17T04:31:42Z   species   
..              ...                  ...                   ...       ...   
523             NaN                  NaN  2022-09-27T21:37:53Z     genus   
525             NaN                  NaN  2020-08-09T23:54:56Z     genus   
530          olivii                  NaN  2019-11-23T00:55:48Z   species   
532      planifolia                  NaN  2019-11-23T05:28:46Z   species   
535             NaN                  NaN  2022-10-17T14:37:53Z     genus   

                                            references  
0    http://www.catalogueoflife.org/annual-checklis...  
2    http://www.catalogueoflife.org/annual-checklis...  
3    http://www.catalogueoflife.org/annual-checklis...  
4    http://www.catalogueoflife.org/annual-checklis...  
6                                                  NaN  

In [98]:
# there's no status but there is a matching inat taxon (id is the taxon id)
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions.sort_values(['scientificName'])
additions['action'] = 'ADD'
additions = additions[['action','scientificName','status_id','id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
additions.columns = additions.columns.str.replace("new_", "", regex=True)
additions = additions.rename(columns={'scientificName':'taxon_name',
                                      'id':'taxon_id',
                                  'status_id':'id'})
additions

action              taxon_name   id taxon_id                 status  \
0      ADD               Acianthus  NaN   202580          Least concern   
2      ADD    Acianthus fornicatus  NaN   202579  Special least concern   
3      ADD     Acianthus ledwardii  NaN  1239753  Special least concern   
4      ADD    Acriopsis emarginata  NaN  1141144             Vulnerable   
6      ADD  Alsophila cunninghamii  NaN   407141        Near Threatened   
..     ...                     ...  ...      ...                    ...   
523    ADD            Thrixspermum  NaN   140357           Confidential   
525    ADD                Trachoma  NaN  1069342           Confidential   
530    ADD           Turnix olivii  NaN    20902  Critically Endangered   
532    ADD      Vanilla planifolia  NaN    61393           Confidential   
535    ADD                 Zeuxine  NaN   124430           Confidential   

           iucn_equivalent                                  authority  \
0            Least Concern  Qld Department of Environment and Science   
2            Least Concern  Qld Department of Environment and Science   
3            Least Concern  Qld Department of Environment and Science   
4               Vulnerable  Qld Department of Environment and Science   
6          Near Threatened  Qld Department of Environment and Science   
..                     ...                                        ...   
523             Vulnerable  Qld Department of Environment and Science   
525             Vulnerable  Qld Department of Environment and Science   
530  Critically Endangered  Qld Department of Environment and Science   
532             Vulnerable  Qld Department of Environment and Science   
535             Vulnerable  Qld Department of Environment and Science   

                                                   url geoprivacy place_id  \
0    https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
2    https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
3    https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
4    https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
6    https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
..                                                 ...        ...      ...   
523  https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
525  https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
530  https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
532  https://apps.des.qld.gov.au/species-search/det...   obscured     7308   
535  https://apps.des.qld.gov.au/species-search/det...   obscured     7308   

      username                                        description  
0    peggydnew  Listed as Confidential - refer to https://www....  
2    peggydnew  Listed as Confidential - refer to https://www....  
3    peggydnew  Listed as Confidential - refer to https://www....  
4    peggydnew  Listed as Confidential - refer to https://www....  
6    peggydnew  Listed as Confidential - refer to https://www....  
..         ...                                                ...  
523  peggydnew  Listed as Confidential - refer to https://www....  
525  peggydnew  Listed as Confidential - refer to https://www....  
530  peggydnew  Listed as Confidential - refer to https://www....  
532  peggydnew  Listed as Confidential - refer to https://www....  
535  peggydnew  Listed as Confidential - refer to https://www....  

[278 rows x 12 columns]

In [99]:
# write these to the file
pd.concat([updates,additions]).to_csv(sourcedir + "qld.csv", index=False)

In [50]:
# what didnt match to a taxon?
unknownToInat = noinatstatus[additions['id'].isna()]
unknownToInat

wildnetTaxonID                  scientificName                 status  \
1              7974              Acianthus borealis  Special least concern   
5             40762             Alsophila baileyana          Least concern   
7             41582                Alsophila exilis             Endangered   
21            30917             Aponogeton cuneatus  Special least concern   
22            21902  Aponogeton elongatus elongatus        Near Threatened   
..              ...                             ...                    ...   
529           11357     Trichoglottis australiensis             Vulnerable   
531           40770                  Vanda whiteana          Least concern   
533           30794               Vrydagzynea grayi             Endangered   
534           22659               Wollumbinia belli             Vulnerable   
536           40091               Zeuxine attenuata          Least concern   

    status_id taxon_id user_id authority status_inat geoprivacy place_id  ...  \
1         NaN      NaN     NaN       NaN         NaN        NaN      NaN  ...   
5         NaN      NaN     NaN       NaN         NaN        NaN      NaN  ...   
7         NaN      NaN     NaN       NaN         NaN        NaN      NaN  ...   
21        NaN      NaN     NaN       NaN         NaN        NaN      NaN  ...   
22        NaN      NaN     NaN       NaN         NaN        NaN      NaN  ...   
..        ...      ...     ...       ...         ...        ...      ...  ...   
529       NaN      NaN     NaN       NaN         NaN        NaN      NaN  ...   
531       NaN      NaN     NaN       NaN         NaN        NaN      NaN  ...   
533       NaN      NaN     NaN       NaN         NaN        NaN      NaN  ...   
534       NaN      NaN     NaN       NaN         NaN        NaN      NaN  ...   
536       NaN      NaN     NaN       NaN         NaN        NaN      NaN  ...   

    phylum class order family genus specificEpithet infraspecificEpithet  \
1      NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
5      NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
7      NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
21     NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
22     NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
..     ...   ...   ...    ...   ...             ...                  ...   
529    NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
531    NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
533    NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
534    NaN   NaN   NaN    NaN   NaN             NaN                  NaN   
536    NaN   NaN   NaN    NaN   NaN             NaN                  NaN   

    modified taxonRank references  
1        NaN       NaN        NaN  
5        NaN       NaN        NaN  
7        NaN       NaN        NaN  
21       NaN       NaN        NaN  
22       NaN       NaN        NaN  
..       ...       ...        ...  
529      NaN       NaN        NaN  
531      NaN       NaN        NaN  
533      NaN       NaN        NaN  
534      NaN       NaN        NaN  
536      NaN       NaN        NaN  

[259 rows x 34 columns]

### are there any that need to be removed?
qld sensitive list count: 955
qld inat statuses count: 653

updates to inat status: 426
additional inat status: 278
qld statuses we can't find a taxon match for in iNaturalist: 259
total: 963 (explainable via the various genus/section entries that we matched to in the taxonomy)

inat statuses left over: 653-426=227 that may need checking

In [95]:
# inat statuses that aren't in added or updated
inatstatuses[~inatstatuses['taxon_id'].isin(updates['taxon_id'])]


status_id taxon_id  user_id place_id source_id  \
408     223608  1015555  3669610     7308             
102     159922  1019990   702203     7308             
667     165360  1023152    58320     7308             
3327    161810  1032816   425992     7308             
263     180953  1033327   702203     7308             
...        ...      ...      ...      ...       ...   
835     262047   977629  3669610     7308             
2615    161807   993333   702203     7308             
2616    161808   993333   702203     6744             
23      167723   993605  3669610     7308             
2897    164338   993923    58320     7308             

                                              authority           status  \
408             Queensland Nature Conservation Act 1992               NT   
102                                          Queensland               VU   
667                               Queensland Government  Near threatened   
3327                              Queensland Government               VU   
263                        Nature Conservation Act 1992               EN   
...                                                 ...              ...   
835             Queensland Nature Conservation Act 1992               NT   
2615                      Nature Conservation Act 1992                VU   
2616  Environment Protection and Biodiversity Conser...               VU   
23                                             QLD DEHP               NT   
2897            Queensland Nature Conservation Act 1992  Near threatened   

                                                    url  \
408   https://apps.des.qld.gov.au/species-search/det...   
102   https://www.legislation.qld.gov.au/view/html/i...   
667   https://apps.des.qld.gov.au/species-search/det...   
3327  https://apps.des.qld.gov.au/species-search/det...   
263   https://bie.ala.org.au/species/https://id.biod...   
...                                                 ...   
835   https://www.data.qld.gov.au/dataset/conservati...   
2615  https://apps.des.qld.gov.au/species-search/det...   
2616  https://apps.des.qld.gov.au/species-search/det...   
23    https://www.data.qld.gov.au/dataset/conservati...   
2897  https://apps.des.qld.gov.au/species-search/det...   

                                       description geoprivacy  ...  \
408                                                            ...   
102                                                            ...   
667                                                            ...   
3327                                                     open  ...   
263                                                            ...   
...                                            ...        ...  ...   
835                                                      open  ...   
2615                                                           ...   
2616                                                           ...   
23                                                   obscured  ...   
2897  See https://www.inaturalist.org/flags/512314       open  ...   

            genus specificEpithet infraspecificEpithet              modified  \
408     Symplocos       harroldii                       2022-11-30T22:01:59Z   
102        Acacia          baueri                       2022-04-06T22:03:46Z   
667        Bertya     pedicellata                       2021-01-05T06:59:04Z   
3327    Macadamia      ternifolia                       2020-08-30T17:12:56Z   
263     Grevillea      linsmithii                       2021-09-20T02:01:57Z   
...           ...             ...                  ...                   ...   
835     Melaleuca         formosa                       2022-09-05T10:47:52Z   
2615  Cupaniopsis      tomentella                       2020-08-30T17:12:00Z   
2616  Cupaniopsis      tomentella                       2020-08-30T17:12:00Z   
23      Acianthus   amplexicaulis                       2021-10-05T08:48:02Z   
2